File: Models/EFDB/Challenge.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using Models.DTO;

namespace Models.EFDB
{
    public class Challenge
    {
        [Key]
        public int Id { get; set; }
        
        public string MaleName { get; set; }
        public string FemaleName { get; set; }
        public string Description { get; set; }

        public string ImageUri { get; set; }

        public int CategoryId { get; set; }

        public string BasicRequirements { get; set; }
        public string ExtraRequirements { get; set; }
 
        public List<ChallengeTask> Tasks { get; set; }
        public DateTime LastModified { get; set; }

        public static explicit operator DTO.Challenge(Challenge challenge)
        {
            var names = new GenderName[]
            {
                new GenderName{Gender = Gender.Male, Name = challenge.MaleName},
                new GenderName{Gender = Gender.Female, Name = challenge.FemaleName}
            };

            var bReqs = challenge.BasicRequirements.Split(',').Select(r => int.Parse(r)).ToArray();
            var basicRequirements = new ChallengeRequirements
            {
                Cubs = bReqs[0],
                Scouts = bReqs[1],
                Guides = bReqs[2],
            };
   
            var eReqs = challenge.ExtraRequirements.Split(',').Select(r => int.Parse(r)).ToArray();
            var extraRequirements = new ChallengeRequirements
            {
                Cubs = eReqs[0],
                Scouts = eReqs[1],
                Guides = eReqs[2],
            };

            return new DTO.Challenge
            {
                Id = challenge.Id,
                Category = (ChallengeCategoryEnum)challenge.CategoryId,

                Description = challenge.Description,
                Names = names,
                ImageUri = new Image(challenge.ImageUri),

                BasicRequirements = basicRequirements,
                ExtraRequirements = extraRequirements,

                BasicTasks = challenge.Tasks.Where(t => !t.Extra).Select(t => (DTO.ChallengeTask)t).ToList(),
                ExtraTasks = challenge.Tasks.Where(t => t.Extra).Select(t => (DTO.ChallengeTask)t).ToList(),
            };
        }

        public static explicit operator Challenge(DTO.Challenge challenge)
        {
            var createdChallenge = new Challenge
            {
                Id = challenge.Id,
                MaleName = challenge.Names[0].Name,
                FemaleName = challenge.Names[1].Name,
                Description = challenge.Description,

                CategoryId = (int)challenge.Category,
                ImageUri = challenge.ImageUri.Uri,
                BasicRequirements = challenge.BasicRequirements.ToString(),
                ExtraRequirements = challenge.ExtraRequirements.ToString(),

                LastModified = DateTime.UtcNow,
            };

            var bTasks = challenge.BasicTasks.Select(task => new ChallengeTask
            {
                Id = task.Id,
                Challenge = createdChallenge,
                ChallengeId = createdChallenge.Id,
                Name = task.Name,
                Description = task.Description,
                Competences = task.Competences.Aggregate("", (a, s) => a += ";" + s, res => res),
                Extra = false,
            }).ToList();

            var eTasks = challenge.ExtraTasks.Select(task => new ChallengeTask
            {
                Id = task.Id,
                Challenge = createdChallenge,
                ChallengeId = createdChallenge.Id,
                Name = task.Name,
                Description = task.Description,

                Extra = true,
            }).ToList();

            createdChallenge.Tasks = new List<ChallengeTask>(bTasks);
            createdChallenge.Tasks.AddRange(eTasks);

            return createdChallenge;
        }
    }
}

File: reactapp/src/models/Challenge.js
import {Task} from "./Task";

export class Challenge {
    constructor(challenge) {
        this.names = {...challenge.names};
        this.description = challenge.description;
        this.requirements = {...challenge.requirements};
        this.tasks = challenge.tasks;
        this.id = challenge.id;
        this.category = challenge.category;
        this.imageUri = challenge.imageUri;
    }

    static fromDto(dto) {
        const id = dto.Id;
        const category = dto.Category;
        const names = { male: dto.Names[0].Name, female: dto.Names[1].Name };
        const description = dto.Description;
        const requirements = {
            basic: {
                cubs: dto.BasicRequirements.Cubs,
                scouts: dto.BasicRequirements.Scouts,
                guides: dto.BasicRequirements.Guides,
            },
            extra: {
                cubs: dto.ExtraRequirements.Cubs,
                scouts: dto.ExtraRequirements.Scouts,
                guides: dto.ExtraRequirements.Guides,
            },
        };
        const imageUri = dto.ImageUri.Uri;
        const tasks = {
            basic: dto.BasicTasks.map((tDto) => Task.fromDto(tDto)),
            extra: dto.ExtraTasks.map((tDto) => Task.fromDto(tDto)),
        };

        return new Challenge({id, names, description, category, requirements, tasks, imageUri});
    }

    toDto = () => {
        const BasicRequirements = {
            Cubs: this.requirements.basic.cubs,
            Scouts: this.requirements.basic.scouts,
            Guides: this.requirements.basic.guides,
        };
        const ExtraRequirements = {
            Cubs: this.requirements.extra.cubs,
            Scouts: this.requirements.extra.scouts,
            Guides: this.requirements.extra.guides,
        };
        const Category = this.category;
        const Names = [{Name: this.names.male, Gender: 0}, {Name: this.names.female, Gender: 1}];
        const Description = this.description;
        const ImageUri = {Uri: this.imageUri, LocalPath: null};
        const BasicTasks = this.tasks.basic.map((task) => task.toDto());
        const ExtraTasks = this.tasks.extra.map((task) => task.toDto());
        const Id = this.id;
        console.log(ImageUri);
        return {BasicRequirements, ExtraRequirements, Category, Names, Description, ImageUri, BasicTasks, ExtraTasks, Id};
    };
}